bagua-core-internal/src/datatypes/mod.rs

```diff
@@ -1201,6 +1221,49 @@ impl BaguaBucket {
         self.inner.lock().comm_ops.push(comm_op);
     }
 
+    pub fn append_decentralized_asynchronous_op(
+        &mut self,
+        communicator_internode: Option<&BaguaSingleCommunicator>,
+        communicator_intranode: Option<&BaguaSingleCommunicator>,
+        peer_selection_mode: String,
+        torch_stream: u64,
+    ) {
+        let communicator =
+            BaguaCommunicator::new(communicator_internode, communicator_intranode, false)
+                .expect("cannot create communicator");
+
+        let mut dst_ready_event = std::ptr::null_mut() as cudaEvent_t;
+        let mut src_ready_event = std::ptr::null_mut() as cudaEvent_t;
+
+        let dst_ready_event_ptr = &mut dst_ready_event;
+        let src_ready_event_ptr = &mut src_ready_event;
+
+        unsafe {
+            cpp::cpp!([dst_ready_event_ptr as "cudaEvent_t *",
+                      src_ready_event_ptr as "cudaEvent_t *"]
+            {
+                CUDACHECK(cudaEventCreate(dst_ready_event_ptr));
+                CUDACHECK(cudaEventCreate(src_ready_event_ptr));
```

Comment:
Memory leak here. They are never released.

You need to create a pool for events like here: https://github.com/BaguaSys/bagua-core/blob/master/bagua-core-internal/src/resource_pool/mod.rs

and implement the `Drop` trait for the event wrapper

Reply:
done